### PR TITLE
Interpolation routines for tracers with mapped_z.

### PR DESCRIPTION
**Pull request by atmyers**
This code is from @asalmgren and @RevathiJambunathan .

The proposed changes:
- [ ] fix a bug or incorrect behavior in AMReX
- [x] add new capabilities to AMReX
- [ ] changes answers in the test suite to more than roundoff level
- [ ] are likely to significantly affect the results of downstream AMReX users
- [ ] include documentation in the code and/or rst files, if appropriate

### CODE REVIEW

File: Src/Particle/AMReX_TracerParticle_mod_K.H [hunk 1]
```diff
@@ -135,5 +141,263 @@ void linear_interpolate_to_particle (const P& p,
...
+                              +   height_arr(i0 + 1                       , j0 + (!is_nodal[d][1]), k )
+                              +   height_arr(i0 + 1 + (!(is_nodal[d][0])) , j0 + (!is_nodal[d][1]), k ) );
+
+        amrex::Real y4 = 0.25 * ( height_arr(i0 + 1                       , j0 + 1                    , k )
+                              +   height_arr(i0 + 1 + (!(is_nodal[d][0])) , j0 + 1                    , k )
+                              +   height_arr(i0 + 1                       , j0 + 1 + (!is_nodal[d][1]), k )
+                              +   height_arr(i0 + 1 + (!(is_nodal[d][0])) , j0 + 1 + (!is_nodal[d][1]), k ) );
+
+        amrex::Real hint_ilo = (p.pos(1) - y1) / (y2 - y1);
+        amrex::Real hint_ihi = (p.pos(1) - y3) / (y4 - y3);
```
Comment (RevathiJambunathan) [inline Src/Particle/AMReX_TracerParticle_mod_K.H]: int yctr = 0;
        amrex::Real ht[4];
        for (int ii=0; ii < 2; ++ii) {
        for (int jj=0; jj < 2; ++jj) {
            ht[yctr] =  0.25 * ( height_arr(i0 + ii                       , j0 + jj                    , k )
                             +   height_arr(i0 + ii + (!(is_nodal[d][0])) , j0 + jj                    , k )
                             +   height_arr(i0 + ii                       , j0 + jj + (!is_nodal[d][1]), k )
                             +   height_arr(i0 + ii + (!(is_nodal[d][0])) , j0 + jj + (!is_nodal[d][1]), k ) );
            ++yctr;
        }}

        amrex::Real hint_ilo = (p.pos(1) - ht[0]) / (ht[1] - ht[0]);
        amrex::Real hint_ihi = (p.pos(1) - ht[2]) / (ht[3] - ht[2]);

Comment (RevathiJambunathan) [reply Src/Particle/AMReX_TracerParticle_mod_K.H]: @atmyers this can replace lines 283 to 304

File: Src/Particle/AMReX_TracerParticle_mod_K.H [hunk 2]
```diff
@@ -135,5 +141,255 @@ void linear_interpolate_to_particle (const P& p,
...
+    {
+        AMREX_D_TERM(amrex::Real lx = (Real(p.pos(0))-plo[0])*dxi[0] - static_cast<Real>(!is_nodal[d][0])*Real(0.5);,
+                     amrex::Real ly = (Real(p.pos(1))-plo[1])*dxi[1] - static_cast<Real>(!is_nodal[d][1])*Real(0.5);,);
+
+        int const i0 = static_cast<int>(amrex::Math::floor(lx));
+        int k = 0; // temporary for 2D
+
+#if (AMREX_SPACEDIM == 2)
+
+        int const j  = static_cast<int>(amrex::Math::floor(ly));
```
Comment (RevathiJambunathan) [inline Src/Particle/AMReX_TracerParticle_mod_K.H]:
```suggestion
        int const j  = p.idata(0);
```

File: Src/Particle/AMReX_TracerParticle_mod_K.H [hunk 3]
```diff
@@ -135,5 +141,255 @@ void linear_interpolate_to_particle (const P& p,
...
+        }
+        amrex::Real hint_ilo = (p.pos(1) - ht[0]) / (ht[1] - ht[0]);
+        amrex::Real hint_ihi = (p.pos(1) - ht[2]) / (ht[3] - ht[2]);
+
+        amrex::Real sy[] = { amrex::Real(1.) - hint_ilo, amrex::Real(1.) - hint_ihi,
+                                               hint_ilo,                   hint_ihi};
+
+#elif (AMREX_SPACEDIM == 3)
+
+        int const j0 = static_cast<int>(amrex::Math::floor(ly));
```
Comment (RevathiJambunathan) [inline Src/Particle/AMReX_TracerParticle_mod_K.H:306]:
```suggestion
        int const j0 = static_cast<int>(amrex::Math::floor(ly));
        k = p.idata(0);
```

File: Src/Particle/AMReX_TracerParticle_mod_K.H [hunk 4]
```diff
@@ -135,5 +141,256 @@ void linear_interpolate_to_particle (const P& p,
...
+AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE
+void linear_interpolate_to_particle_z (const P& p,
+                                       amrex::GpuArray<amrex::Real,AMREX_SPACEDIM> const& plo,
+                                       amrex::GpuArray<amrex::Real,AMREX_SPACEDIM> const& dxi,
+                                       const Array4<amrex::Real const>* data_arr,
+                                       const amrex::Array4<amrex::Real const>&  height_arr,
+                                       amrex::ParticleReal * val,
+                                       const IntVect* is_nodal,
+                                       int start_comp, int ncomp, int num_arrays)
+{
```
Comment (RevathiJambunathan) [inline Src/Particle/AMReX_TracerParticle_mod_K.H:249]:
```suggestion
{
#if (AMREX_SPACEDIM == 1)
    amrex::ignore_unused(p, plo, dxi, data_arr, height_arr, val, is_nodal, start_comp, ncomp, num_arrays);
    amrex::Abort(" Terrain fitted grid interpolation is not supported in 1D\n");
+#else
```

File: Src/Particle/AMReX_TracerParticle_mod_K.H [hunk 5]
```diff
@@ -135,5 +141,256 @@ void linear_interpolate_to_particle (const P& p,
...
+                        val[ctr] += static_cast<ParticleReal>(
+                                    (data_arr[d])(i0+ii, j0+jj, k0 + kk, comp)*sx[ii]*sy[jj]*sz[sz_ctr]);
+                        ++sz_ctr;
+                    }
+                }
+            }
+#endif
+            ctr++;
+        } // ncomp
+    } // d
```
Comment (RevathiJambunathan) [inline Src/Particle/AMReX_TracerParticle_mod_K.H:395]:
```suggestion
    } // d
#endif
```

File: Src/Particle/AMReX_TracerParticle_mod_K.H (hunk 4 above)
Comment (RevathiJambunathan) [reply Src/Particle/AMReX_TracerParticle_mod_K.H:249]: that would fix the 1D compilation?

File: Src/Particle/AMReX_TracerParticle_mod_K.H [hunk 6]
```diff
@@ -135,5 +141,256 @@ void linear_interpolate_to_particle (const P& p,
...
+    {
+        AMREX_D_TERM(amrex::Real lx = (Real(p.pos(0))-plo[0])*dxi[0] - static_cast<Real>(!is_nodal[d][0])*Real(0.5);,
+                     amrex::Real ly = (Real(p.pos(1))-plo[1])*dxi[1] - static_cast<Real>(!is_nodal[d][1])*Real(0.5);,);
+
+        int const i0 = static_cast<int>(amrex::Math::floor(lx));
+        int k = 0; // temporary for 2D
+
+#if (AMREX_SPACEDIM == 2)
+
+        int const j  = p.idata(0);
```
Comment (RevathiJambunathan) [inline Src/Particle/AMReX_TracerParticle_mod_K.H]:
```suggestion
        amrex::ignored_unused(ly);
        int const j  = p.idata(0);
```

File: Src/Particle/AMReX_TracerParticle_mod_K.H [hunk 7]
```diff
@@ -135,5 +141,256 @@ void linear_interpolate_to_particle (const P& p,
...
+
+        amrex::Real hlo_xlo = 0.25 * ( height_arr(i0                       , j                     , k)
```
Comment (RevathiJambunathan) [inline Src/Particle/AMReX_TracerParticle_mod_K.H]:
```suggestion
        amrex::Real hlo_xlo = amrex::Real(0.25) * ( height_arr(i0                       , j                     , k)
```

File: Src/Particle/AMReX_TracerParticle_mod_K.H [hunk 8]
```diff
@@ -135,5 +141,256 @@ void linear_interpolate_to_particle (const P& p,
...
+        amrex::Real sx[] = { amrex::Real(1.) - xint, xint};
+        amrex::Real height_at_px = sx[0] * hlo_xlo + sx[1] * hlo_xhi;
+
+        int const j0 = (amrex::Real(p.pos(1)) >= height_at_px) ? j : j-1;
+
+        int yctr = 0;
+        amrex::Real ht[4];
+        for (int ii=0; ii < 2; ++ii) {
+            for (int jj=0; jj < 2; ++jj) {
+                ht[yctr] = 0.25 * ( height_arr(i0 + ii                       , j0 + jj , k                    )
```
Comment (RevathiJambunathan) [inline Src/Particle/AMReX_TracerParticle_mod_K.H]:
```suggestion
                ht[yctr] = amrex::Real(0.25) * ( height_arr(i0 + ii                       , j0 + jj , k                    )
```

File: Src/Particle/AMReX_TracerParticle_mod_K.H [hunk 9]
```diff
@@ -135,5 +141,256 @@ void linear_interpolate_to_particle (const P& p,
...
+        amrex::Real const yint = ly - static_cast<Real>(j0);
+        amrex::Real sx[] = { amrex::Real(1.) - xint, xint};
+        amrex::Real sy[] = { amrex::Real(1.) - yint, yint};
+
+        amrex::Real hlo[4];
+        int ilo = 0;
+        amrex::Real height_at_pxy = 0.;
+        for (int ii = 0; ii < 2; ++ii) {
+            for (int jj = 0; jj < 2; ++jj) {
+                hlo[ilo] = 0.125
```
Comment (RevathiJambunathan) [inline Src/Particle/AMReX_TracerParticle_mod_K.H]:
```suggestion
                hlo[ilo] = amrex::Real(0.125)
```

File: Src/Particle/AMReX_TracerParticle_mod_K.H [hunk 10]
```diff
@@ -135,5 +141,256 @@ void linear_interpolate_to_particle (const P& p,
...
+
+        int const k0 = (amrex::Real(p.pos(2)) >= height_at_pxy ) ? k : k-1;
+
+        // to make z0-z7 concise
+        int zctr = 0;
+        amrex::Real ht[8];
+        for (int ii = 0; ii < 2; ++ii) {
+        for (int jj = 0; jj < 2; ++jj) {
+        for (int kk = 0; kk < 2; ++kk) {
+            ht[zctr] = 0.125 *
```
Comment (RevathiJambunathan) [inline Src/Particle/AMReX_TracerParticle_mod_K.H]:
```suggestion
            ht[zctr] = amrex::Real(0.125) *
```

File: Src/Particle/AMReX_TracerParticle_mod_K.H [hunk 11]
```diff
@@ -135,5 +141,261 @@ void linear_interpolate_to_particle (const P& p,
...
+                                       const Array4<amrex::Real const>* data_arr,
+                                       const amrex::Array4<amrex::Real const>&  height_arr,
+                                       amrex::ParticleReal * val,
+                                       const IntVect* is_nodal,
+                                       int start_comp, int ncomp, int num_arrays)
+{
+#if (AMREX_SPACEDIM == 1)
+    amrex::ignore_unused(p, plo, dxi, data_arr, height_arr, val, is_nodal, start_comp, ncomp, num_arrays);
+    amrex::Abort(" Terrain fitted grid interpolation is not supported in 1D\n");
++#else
```
Comment (RevathiJambunathan) [inline Src/Particle/AMReX_TracerParticle_mod_K.H]:
```suggestion
#else
```

**Review by RevathiJambunathan — APPROVED** [review]

File: Src/Particle/AMReX_TracerParticle_mod_K.H [hunk 12]
```diff
@@ -135,5 +141,261 @@ void linear_interpolate_to_particle (const P& p,
...
+    for (int d = 0; d < num_arrays; d++)
+    {
+        AMREX_D_TERM(amrex::Real lx = (Real(p.pos(0))-plo[0])*dxi[0] - static_cast<Real>(!is_nodal[d][0])*Real(0.5);,
+                     amrex::Real ly = (Real(p.pos(1))-plo[1])*dxi[1] - static_cast<Real>(!is_nodal[d][1])*Real(0.5);,);
+
+        int const i0 = static_cast<int>(amrex::Math::floor(lx));
+        int k = 0; // temporary for 2D
+
+#if (AMREX_SPACEDIM == 2)
+        amrex::ignored_unused(ly);
```
Comment (atmyers) [inline Src/Particle/AMReX_TracerParticle_mod_K.H]:
```suggestion
        amrex::ignore_unused(ly);
```

**Review by asalmgren — APPROVED** [review]

**Pull request merged**